cpp/src/parquet/file_reader.cc

```diff
@@ -373,6 +373,7 @@ class SerializedFile : public ParquetFileReader::Contents {
 
   void set_metadata(std::shared_ptr<FileMetaData> metadata) {
     file_metadata_ = std::move(metadata);
+    file_decryptor_ = file_metadata_->file_decryptor();
```

Comment:
This is confusing, here we set `file_decryptor_` from `file_metadata_`, but in `ParseUnencryptedFileMetadata` and `ParseMetaDataOfEncryptedFileWithPlaintextFooter` we set `file_metadata_` from `file_decryptor_`. Can we please make this consistent?

Reply:
https://github.com/apache/arrow/blob/main/cpp/src/parquet/file_reader.cc#L775-L780

I think these are in two different directions while opening a parquet reader: 
- For `ParseXXX` functions where we parse footer to create `file_decryptor_` and `file_metadata_`. We need to set `file_decryptor_` to `file_metadata_` so that both `SerializedFile` and `FileMetaData` have a copy of the decryptor.
- For `set_metadata` function, we already have the cached `FileMetaData` but need to create `SerializedFile` where its `file_decryptor_` is null. Therefore we need to get `file_decryptor_` from `file_metadata_`.

Does that make sense?

Reply:
I have removed `file_decryptor_` from `SerializedFile` and `SerializedRowGroup`. Let me know WDYT. @pitrou